message/validation/consensus_validation.go

```diff
@@ -0,0 +1,361 @@
+			// they'd have to wait for the next slot/round to be accepted.
+			err := ErrRoundAlreadyAdvanced
+			err.want = signerState.Round
+			err.got = msgRound
+			return err
+		}
+
+		if !(msgSlot > signerState.Slot || msgSlot == signerState.Slot && msgRound > signerState.Round) {
+			if err := mv.validateDutiesCount(signerState, msg.MsgID.GetRoleType()); err != nil {
+				return err
+			}
+
+			if mv.hasFullData(signedMsg) && signerState.ProposalData != nil && !bytes.Equal(signerState.ProposalData, signedMsg.FullData) {
+				return ErrDuplicatedProposalWithDifferentData
+			}
+
+			limits := maxMessageCounts(len(share.Committee), int(share.Quorum))
+			if err := signerState.MessageCounts.Validate(msg, limits); err != nil {
+				return err
+			}
+		}
+	}
+
+	if err := mv.validateJustifications(share, signedMsg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) validateDutiesCount(state *SignerState, role spectypes.BeaconRole) error {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator, spectypes.BNRoleValidatorRegistration:
+		if state.EpochDuties > maxDutiesPerEpoch {
+			err := ErrTooManyDutiesPerEpoch
+			err.got = fmt.Sprintf("%v (role %v)", state.EpochDuties, role)
+			err.want = maxDutiesPerEpoch
+			return err
+		}
+		// TODO: check other roles
+	}
+
+	return nil
+}
+
+func (mv *MessageValidator) hasFullData(signedMsg *specqbft.SignedMessage) bool {
+	return (signedMsg.Message.MsgType == specqbft.ProposalMsgType ||
+		signedMsg.Message.MsgType == specqbft.RoundChangeMsgType ||
+		mv.isDecidedMessage(signedMsg)) && len(signedMsg.FullData) != 0 // TODO: more complex check of FullData
+}
+
+func (mv *MessageValidator) isDecidedMessage(signedMsg *specqbft.SignedMessage) bool {
+	return signedMsg.Message.MsgType == specqbft.CommitMsgType && len(signedMsg.Signers) > 1
+}
+
+func (mv *MessageValidator) maxRound(role spectypes.BeaconRole) specqbft.Round {
+	switch role {
+	case spectypes.BNRoleAttester, spectypes.BNRoleAggregator: // TODO: check if value for aggregator is correct as there are messages on stage exceeding the limit
+		return 12 // TODO: consider calculating based on quick timeout and slow timeout
+	case spectypes.BNRoleProposer, spectypes.BNRoleSyncCommittee, spectypes.BNRoleSyncCommitteeContribution:
+		return 6
+	case spectypes.BNRoleValidatorRegistration:
+		return 0
+	default:
+		panic("unknown role")
+	}
+}
+
+func (mv *MessageValidator) currentEstimatedRound(sinceSlotStart time.Duration) specqbft.Round {
+	if currentQuickRound := specqbft.FirstRound + specqbft.Round(sinceSlotStart/roundtimer.QuickTimeout); currentQuickRound <= roundtimer.QuickTimeoutThreshold {
+		return currentQuickRound
+	}
+
+	sinceFirstSlowRound := sinceSlotStart - (time.Duration(roundtimer.QuickTimeoutThreshold) * roundtimer.QuickTimeout)
+	estimatedRound := roundtimer.QuickTimeoutThreshold + specqbft.FirstRound + specqbft.Round(sinceFirstSlowRound/roundtimer.SlowTimeout)
+	return estimatedRound
+}
```

Comment:
same as `waitAfterSlotStart`, unless you think it will be used outside of test?

Reply:
same as waitAfterSlotStart